src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,672 @@
+
+  return ((m_state == STATE_SHUTDOWN) ||
+          (!m_actions_contexts.empty() &&
+           m_actions_contexts.back().first == ACTION_SHUT_DOWN));
+}
+
+template <typename I>
+void ManagedLock<I>::send_acquire_lock() {
+  assert(m_lock.is_locked());
+  if (m_state == STATE_LOCKED) {
+    complete_active_action(STATE_LOCKED, 0);
+    return;
+  }
+
+  ldout(m_cct, 10) << dendl;
+  m_state = STATE_ACQUIRING;
+
+  uint64_t watch_handle = m_watcher->get_watch_handle();
+  if (watch_handle == 0) {
+    lderr(m_cct) << "watcher not registered - delaying request" << dendl;
+    m_state = STATE_WAITING_FOR_REGISTER;
+    return;
+  }
+  m_cookie = ManagedLock<I>::encode_lock_cookie(watch_handle);
+
+  pre_acquire_lock_handler(util::create_context_callback<
+      ManagedLock<I>, &ManagedLock<I>::handle_pre_acquire_lock>(this));
+}
+
+template <typename I>
+void ManagedLock<I>::handle_pre_acquire_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  if (r < 0) {
+    handle_acquire_lock(r);
+    return;
+  }
+
+  using managed_lock::AcquireRequest;
+  AcquireRequest<I>* req = AcquireRequest<I>::create(m_ioctx, m_watcher,
+      m_work_queue, m_oid, m_cookie,
+      util::create_context_callback<
+        ManagedLock<I>, &ManagedLock<I>::handle_acquire_lock>(this));
+  m_work_queue->queue(new C_SendLockRequest<AcquireRequest<I>>(req), 0);
+}
+
+template <typename I>
+void ManagedLock<I>::handle_acquire_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  if (r == -EBUSY || r == -EAGAIN) {
+    ldout(m_cct, 5) << ": unable to acquire exclusive lock" << dendl;
+  } else if (r < 0) {
+    lderr(m_cct) << ": failed to acquire exclusive lock:" << cpp_strerror(r)
+               << dendl;
+  } else {
+    ldout(m_cct, 5) << ": successfully acquired exclusive lock" << dendl;
+  }
+
+  m_post_next_state = (r < 0 ? STATE_UNLOCKED : STATE_LOCKED);
+
+  Mutex::Locker locker(m_lock);
+  post_acquire_lock_handler(r, util::create_context_callback<
+      ManagedLock<I>, &ManagedLock<I>::handle_post_acquire_lock>(this));
+}
+
+template <typename I>
+void ManagedLock<I>::handle_post_acquire_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (r < 0 && m_post_next_state == STATE_LOCKED) {
+    // release_lock without calling pre and post callbacks
+    revert_to_unlock_state(r);
+  } else {
+    complete_active_action(m_post_next_state, r);
+  }
+}
+
+template <typename I>
+void ManagedLock<I>::revert_to_unlock_state(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  using managed_lock::ReleaseRequest;
+  ReleaseRequest<I>* req = ReleaseRequest<I>::create(m_ioctx, m_watcher,
+      m_work_queue, m_oid, m_cookie,
+      new FunctionContext([this, r](int ret) {
+        Mutex::Locker locker(m_lock);
+        if (ret >= 0) {
+          complete_active_action(STATE_UNLOCKED, r);
+        } else {
+          ldout(m_cct, 10) << ": failed to revert to unlocked state:"
```

Comment:
What are the valid failure cases here? If we get ENOENT / EBLACKLISTED, I think it's safe to assume we are unlocked (probably could mask those errors within the ReleaseRequest state machine to simplify things). Otherwise, I think we should still treat an error as we are unlocked -- especially since the derived class has already thrown an error so it isn't set up for a locked state.